File: src/backend/domain/entities/ImageJob.ts
export type JobStatus = 'pending' | 'processing' | 'completed' | 'failed' | 'cancelled';

export interface ImageJobEntity {
  id: string;
  originalPath: string;
  originalName: string;
  mimeType: string;
  fileSize: number;
  prompt: string;
  status: JobStatus;
  outputPath?: string;
  errorMessage?: string;
  retryCount: number;
  createdAt: Date;
  updatedAt: Date;
  processingStartedAt?: Date;
  processingCompletedAt?: Date;
  processingDurationMs?: number;
}

export class ImageJob implements ImageJobEntity {
  id: string;
  originalPath: string;
  originalName: string;
  mimeType: string;
  fileSize: number;
  prompt: string;
  status: JobStatus;
  outputPath?: string;
  errorMessage?: string;
  retryCount: number;
  createdAt: Date;
  updatedAt: Date;
  processingStartedAt?: Date;
  processingCompletedAt?: Date;
  processingDurationMs?: number;

  constructor(data: ImageJobEntity) {
    this.id = data.id;
    this.originalPath = data.originalPath;
    this.originalName = data.originalName;
    this.mimeType = data.mimeType;
    this.fileSize = data.fileSize;
    this.prompt = data.prompt;
    this.status = data.status;
    this.outputPath = data.outputPath;
    this.errorMessage = data.errorMessage;
    this.retryCount = data.retryCount;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
    this.processingStartedAt = data.processingStartedAt;
    this.processingCompletedAt = data.processingCompletedAt;
    this.processingDurationMs = data.processingDurationMs;
  }

  markAsProcessing(): void {
    this.status = 'processing';
    this.processingStartedAt = new Date();
    this.updatedAt = new Date();
  }

  markAsCompleted(outputPath: string): void {
    this.status = 'completed';
    this.outputPath = outputPath;
    this.processingCompletedAt = new Date();
    if (this.processingStartedAt) {
      this.processingDurationMs = this.processingCompletedAt.getTime() - this.processingStartedAt.getTime();
    }
    this.updatedAt = new Date();
  }

  markAsFailed(errorMessage: string): void {
    this.status = 'failed';
    this.errorMessage = errorMessage;
    this.updatedAt = new Date();
  }

  markAsCancelled(): void {
    this.status = 'cancelled';
    this.updatedAt = new Date();
  }

  incrementRetry(): void {
    this.retryCount++;
    this.updatedAt = new Date();
  }

  canRetry(maxRetries: number): boolean {
    return this.retryCount < maxRetries;
  }
}

File: app/domain/entities/image_job.py
"""ImageJob domain entity.

Python equivalent of `src/backend/domain/entities/ImageJob.ts`.
Keeps the same lifecycle methods so the queue service reads identically.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Literal, Optional

JobStatus = Literal["pending", "processing", "completed", "failed", "cancelled"]


def _now() -> datetime:
    return datetime.now(timezone.utc)


@dataclass
class ImageJob:
    id: str
    original_path: str
    original_name: str
    mime_type: str
    file_size: int
    prompt: str
    status: JobStatus = "pending"
    output_path: Optional[str] = None
    error_message: Optional[str] = None
    retry_count: int = 0
    created_at: datetime = field(default_factory=_now)
    updated_at: datetime = field(default_factory=_now)
    processing_started_at: Optional[datetime] = None
    processing_completed_at: Optional[datetime] = None
    processing_duration_ms: Optional[int] = None

    # ── State transitions ──────────────────────────────────────────────────
    def mark_as_processing(self) -> None:
        self.status = "processing"
        self.processing_started_at = _now()
        self.updated_at = _now()

    def mark_as_completed(self, output_path: str) -> None:
        self.status = "completed"
        self.output_path = output_path
        self.processing_completed_at = _now()
        if self.processing_started_at:
            delta = self.processing_completed_at - self.processing_started_at
            self.processing_duration_ms = int(delta.total_seconds() * 1000)
        self.updated_at = _now()

    def mark_as_failed(self, error_message: str) -> None:
        self.status = "failed"
        self.error_message = error_message
        self.updated_at = _now()

    def mark_as_cancelled(self) -> None:
        self.status = "cancelled"
        self.updated_at = _now()

    def increment_retry(self) -> None:
        self.retry_count += 1
        self.updated_at = _now()

    def can_retry(self, max_retries: int) -> bool:
        return self.retry_count < max_retries
